src/policykitagent.cpp

```diff
@@ -116,27 +116,36 @@ void PolicykitAgent::request(const QString &request, bool echo)
 
     PolkitQt1::Identity identity = m_SessionIdentity[session];
     m_gui->setPrompt(identity, request, echo);
-    if (m_gui->exec())
-        session->setResponse(m_gui->response());
-    else
-        session->setResponse("");
+    connect(m_gui, &QDialog::finished, [this, session] (int result)
+    {
+        if (result == QDialog::Accepted && m_gui->identity().toString() == m_SessionIdentity[session].toString())
+            session->setResponse(m_gui->response());
+        else
+            session->cancel();
+    });
+    m_gui->show();
```

Comment:
No, it can't be blocking... as there can be multiple sessions (identities) in progress. And each of it must react on the finished().